### ISSUE TITLE
Failed gateway withdraw: The generated transaction would be rejected by the federation for being too large

### ISSUE THREAD

**Issue by douglaz** [fedimint/fedimint#5562]
```
$ gateway-cli withdraw --federation-id xxx --amount all --address bc1aaaa
Error: Bad status returned 500 Internal Server Error


$ journalctl -u gatewayd -n 1000|grep ERROR
Jul 02 16:02:51 ggjdy0dfnspwkcn3ddz6b1iycy53qvh6-gatewayd[894]: 2024-07-02T16:02:51.960952Z ERROR withdraw{payload=WithdrawPayload { federation_id: FederationId(xxx), amount: All, address: bc1aaaa }}: ln_gateway::rpc::rpc_server: error=Other: The generated transaction would be rejected by the federation for being too large.
```

**Comment by douglaz** [comment]
Still happens

**Comment by elsirion** [comment]
Even after a few other transactions on 0.4? That would mean the consolidation logic isn't kicking in or isn't aggressive enough @dpc. Maybe we need to fine tune the hysteresis function a bit, it could be keeping too many notes per denomination around.

**Comment by douglaz** [comment]
This is still 0.3

**Comment by elsirion** [comment]
The fix was in #4922, but the backport never happened, iirc because 0.4 was so close and there were some bigger conflicts.
![Screenshot 2024-08-19 at 11-41-20 feat(client) consolidate excessive amount of notes in a given tier by dpc · Pull Request #4922 · fedimint_fedimint](https://github.com/user-attachments/assets/574a78a1-5604-467b-9e3d-5cc97f937cb8)

**Comment by Kodylow** [comment]
TODO: validate this, fixed in 0.4, if working just close

**Comment by m1sterc001guy** [comment]
So I just hit this on a gateway that is running > 0.4

![image](https://github.com/user-attachments/assets/e1d6eb28-90ab-4e32-9044-bf572f359798)

This gateway was recently updated from 0.3.3 -> 0.4.3 -> recent master commit. Then I tried to spend some ecash notes from the gateway and it failed. It is possible that since no transactions occurred after the updates, it didn't have an opportunity to consolidate the notes.

The `spend-ecash` command feels like a footgun though because it is not checking the size of the serialized note produced, it can easily serialize ecash that will be rejected by consensus. I think an easy improvement would be to add checks to `spend-ecash` so that it doesn't produce a serialized note that is larger than the consensus maximum.

**Comment by m1sterc001guy** [comment]
A gateway endpoint that forces the gateway to consolidate some notes might also be useful.

**Comment by elsirion** [comment]
Agree that we should handle receiving too large note bags better. 

@m1sterc001guy for your current problem, this is how you split the notes into smaller ones:

```bash
fmcli dev decode notes "<large_notes_bag>" | jq -c '.federation_id_prefix as $prefix | .notes | to_entries[] | {"federation_id_prefix": $prefix, "notes":  [.] | from_entries}' | while read NOTE; do fmcli dev encode notes "$NOTE"; done
```

**Comment by Kodylow** [comment]
@m1sterc001guy is this something that might fuck up a database upgrade on a gateway to the single db?

**Comment by m1sterc001guy** [comment]
> is this something that might fuck up a database upgrade on a gateway to the single db?

This was how I ran into this issue. I wanted to deplete my balance so I could remove the DB and upgrade to using a client derived from the mnemonic. So I spent my entire balance, then tried to reissue it once I re-joined the federation. 

Was able to get around it by writing a simple python script to parse the notes from the `fedimint-cli split` command. Just wanted to document some potential improvements.
